### PR TITLE
Refactor withdraw

### PR DESCRIPTION
**Pull request by zajck**
Major changes:

introduced new internal function reduceEscrowAmountDeposits to reduce code duplication
introduced distributeAndWithdraw which does similar to what withdraw did before, except it does it only for specified entity
added external function withdrawSingle which allows withdrawing for specific entity
withdraw still exists and it's behaviour is exactly the same as before the changes. The biggest difference is that if this fails, there always exist

Old test are passing (with just minor modifications).
I added new test file `3_1_withdrawals_per_entity` which performs similar actions as the existing `3_withdrawals` except separate calls are made for each entity.
It extends the running time for around 20%.

### CODE REVIEW

**Review by bidyut-arianelabs — CHANGES_REQUESTED** [review]

**Review by hswopeams — DISMISSED** [review]

**Review by hswopeams — DISMISSED** [review]

File: contracts/VoucherKernel.sol [hunk 1]
```diff
@@ -720,16 +720,49 @@ contract VoucherKernel is IVoucherKernel, Ownable, Pausable, ReentrancyGuard {
...
-        vouchersStatus[_tokenIdVoucher].isDepositsReleased = true;
 
-        emit LogFundsReleased(_tokenIdVoucher, 1);
+        // if (_to == Entity.ISSUER) {
+        //     vouchersStatus[_tokenIdVoucher].depositReleased.issuer = true;
+        // } else if (_to == Entity.HOLDER) {
+        //     vouchersStatus[_tokenIdVoucher].depositReleased.holder = true;
+        // } else {
+        //     vouchersStatus[_tokenIdVoucher].depositReleased.pool = true;
+        // }
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/VoucherKernel.sol]: Removed

File: contracts/VoucherKernel.sol [hunk 2]
```diff
@@ -720,16 +720,49 @@ contract VoucherKernel is IVoucherKernel, Ownable, Pausable, ReentrancyGuard {
...
+     * @param _to               recipient, one of {ISSUER, HOLDER, POOL}
+     */
+    function isDepositReleased(uint256 _tokenIdVoucher, Entity _to) external view override returns (bool){
+        // if (_to == Entity.ISSUER) {
+        //     return vouchersStatus[_tokenIdVoucher].depositReleased.issuer;
+        // } else if (_to == Entity.HOLDER) {
+        //     return vouchersStatus[_tokenIdVoucher].depositReleased.holder;
+        // } else {
+        //     return vouchersStatus[_tokenIdVoucher].depositReleased.pool;
+        // }
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/VoucherKernel.sol]: Removed

File: contracts/VoucherKernel.sol [hunk 3]
```diff
@@ -1011,6 +1044,17 @@ contract VoucherKernel is IVoucherKernel, Ownable, Pausable, ReentrancyGuard {
...
 
+    function getTotalDeposits(uint256 _tokenIdSupply)
+        internal
+        view
+        returns (
+            uint256
+        )
+    {
+        bytes32 promiseKey = ordersPromise[_tokenIdSupply];
+        return promises[promiseKey].depositSe +  promises[promiseKey].depositBu;
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/VoucherKernel.sol]: Fixed

**Review by bidyut-arianelabs — DISMISSED** [review]
Looks good to me 👍

**Review by bidyut-arianelabs — APPROVED** [review]

**Review by hswopeams — APPROVED** [review]

**Pull request merged**